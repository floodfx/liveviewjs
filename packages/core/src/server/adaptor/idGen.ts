/**
 * Type that defines a function that returns a string ID used to identify a unique http request
 * and/or websocket connection.  Should generate unique IDs for each request and connection.  Good
 * concrete implementations are: nanoid, shortid, uuidv4 (though these are long).
 */
export type IdGenerator = () => string;
