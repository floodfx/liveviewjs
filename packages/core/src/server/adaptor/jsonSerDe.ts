import { SerDe } from "./serDe";

/**
 * A SerDe (serializer/deserializer) that uses JSON.stringify and JSON.parse.
 * WARNING: this is not secure so should only be used for testing.
 */
export class JsonSerDe<T> implements SerDe<T, string> {
  serialize<T>(obj: T) {
    return Promise.resolve(JSON.stringify(obj));
  }

  deserialize<T>(data: string) {
    return Promise.resolve(JSON.parse(data) as T);
  }
}
