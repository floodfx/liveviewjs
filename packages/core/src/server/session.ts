/**
 * Generically represent session with `string` keys to `any` value and a named member called
 * `_csrf_token` that is used to protect against cross-site request forgery attacks.
 */
export type SessionData = {
  /**
   * The CSRF token used to protect against cross-site request forgery attacks.
   */
  _csrf_token?: string;
  [key: string]: any;
};
