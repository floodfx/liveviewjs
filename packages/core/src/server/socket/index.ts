export * from "./liveSocket";
export * from "./liveViewManager";
export * from "./ws/wsHandler";
export * from "./wsMessageRouter";
