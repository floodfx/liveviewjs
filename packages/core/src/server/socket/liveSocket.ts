import { FileSystemAdaptor } from "../adaptor";
import { AnyLiveContext, AnyLiveInfo, AnyLivePushEvent, LiveContext, LiveInfo } from "../live";
import { UploadConfig, UploadEntry } from "../upload";
import { UploadConfigOptions } from "../upload/uploadConfig";
import { maybeAddStructuredClone } from "./structuredClone";
maybeAddStructuredClone();

/**
 *  Type that enables Info events to be passed as plain strings
 */
export type Info<TInfo extends LiveInfo> = TInfo["type"] | TInfo;

/**
 * The type passed to the `consumeUploadedEntries` callback's "meta" parameter
 */
export type ConsumeUploadedEntriesMeta = {
  /**
   * The location of the file on the server
   */
  path: string;
  /**
   * The FileSystemAdaptor for the platform (node or deno)
   */
  fileSystem: FileSystemAdaptor;
};

/**
 * Main interface to update state, interact, message, and otherwise
 * manage the lifecycle of a `LiveView`.
 *
 * The `LiveView` API (i.e. `mount`, `handleParams`, `handleInfo`, `handleEvent`)
 * are all passed `LiveViewSocket` which provide access to the current `LiveView`
 * context (via `context`) as well as various methods update the `LiveView` including
 * `assign` which updates the `LiveView`'s context (i.e. state).
 */
export interface LiveViewSocket<TContext extends LiveContext = AnyLiveContext, TInfos extends LiveInfo = AnyLiveInfo> {
  /**
   * The id of the `LiveView`
   */
  readonly id: string;
  /**
   * Whether the websocket is connected.
   * true if connected to a websocket, false for http request
   */
  readonly connected: boolean;
  /**
   * The current context (i.e. state) of the `LiveView`
   */
  readonly context: TContext;
  /**
   * `assign` is used to update the context (i.e. state) of the `LiveComponent`
   * @param context a `Partial` of the LiveView's context to update
   */
  assign(context: Partial<TContext>): void;
  /**
   * Marks any set properties as temporary and will be reset to the given
   * value after the next render cycle.  Typically used to ensure large but
   * infrequently updated values are not kept in memory.
   *
   * @param context a partial of the context that should be temporary and the value to reset it to
   */
  tempAssign(context: Partial<TContext>): void;
  /**
   * Updates the `<title>` tag of the `LiveView` page.  Requires using the
   * `live_title` helper in rendering the page.
   *
   * @param newPageTitle the new text value of the page - note the prefix and suffix will not be changed
   */
  pageTitle(newPageTitle: string): void;
  /**
   * Pushes and event (possibly with data) from the server to the client.  Requires
   * either a `window.addEventListener` defined for that event or a client `Hook`
   * to be defined and to be listening for the event via `this.handleEvent` callback.
   *
   * @param pushEvent the event to push to the client
   */
  pushEvent(pushEvent: AnyLivePushEvent): void;
  /**
   * Updates the LiveView's browser URL with the given path and query parameters.
   *
   * @param path the path whose query params are being updated
   * @param params the query params to update the path with
   * @param replaceHistory whether to replace the current history entry or push a new one (defaults to false)
   */
  pushPatch(path: string, params?: URLSearchParams, replaceHistory?: boolean): Promise<void>;
  /**
   * Shutdowns the current `LiveView`and loads another `LiveView`in its place
   * without reloading the whole page (i.e. making a full HTTP request).  Can be
   * used to remount the current `LiveView`if need be. Use `pushPatch` to update the
   * current `LiveView`without unloading and remounting.
   *
   * @param path the path whose query params are being updated
   * @param params the query params to update the path with
   * @param replaceHistory whether to replace the current history entry or push a new one (defaults to false)
   */
  pushRedirect(path: string, params?: URLSearchParams, replaceHistory?: boolean): Promise<void>;
  /**
   * Add flash to the socket for a given key and value.
   * @param key the key to add the flash to
   * @param value the flash value
   */
  putFlash(key: string, value: string): Promise<void>;
  /**
   * Runs the given function on the given interval until this `LiveView` is
   * unloaded.
   *
   * @param fn the function to run on the interval
   * @param intervalMillis the interval to run the function on in milliseconds
   */
  repeat(fn: () => void, intervalMillis: number): void;
  /**
   * Send an internal event (a.k.a "Info") to the LiveView's `handleInfo` method
   *
   * @param event the event to send to `handleInfo`
   */
  sendInfo(info: Info<TInfos>): void;
  /**
   * Subscribe to the given topic using pub/sub. Events published to this topic
   * will be delivered to `handleInfo`.
   *
   * @param topic the topic to subscribe this `LiveView`to
   */
  subscribe(topic: string): Promise<void>;

  /**
   * Allows file uploads for the given `LiveView`and configures the upload
   * options (filetypes, size, etc).
   * @param name the name of the upload
   * @param options the options for the upload (optional)
   */
  allowUpload(name: string, options?: UploadConfigOptions): Promise<void>;

  /**
   * Cancels the file upload for a given UploadConfig by config name and file ref.
   * @param name the name of the upload from which to cancel
   * @param ref the ref of the upload entry to cancel
   */
  cancelUpload(configName: string, ref: string): Promise<void>;

  /**
   * Consume the uploaded files for a given UploadConfig (by name). This
   * should only be called after the form's "save" event has occurred which
   * guarantees all the files for the upload have been fully uploaded.
   * @param name the name of the upload from which to consume
   * @param fn the callback to run for each entry
   * @returns an array of promises based on the return type of the callback function
   * @throws if any of the entries are not fully uploaded (i.e. completed)
   */
  consumeUploadedEntries<T>(
    configName: string,
    fn: (meta: ConsumeUploadedEntriesMeta, entry: UploadEntry) => Promise<T>
  ): Promise<T[]>;

  /**
   * Returns two sets of files that are being uploaded, those `completed` and
   * those `inProgress` for a given UploadConfig (by name).  Unlike `consumeUploadedEntries`,
   * this does not require the form's "save" event to have occurred and will not
   * throw if any of the entries are not fully uploaded.
   * @param name the name of the upload from which to get the entries
   * @returns an object with `completed` and `inProgress` entries
   */
  uploadedEntries(configName: string): Promise<{
    completed: UploadEntry[];
    inProgress: UploadEntry[];
  }>;
}

abstract class BaseLiveViewSocket<TContext extends LiveContext = AnyLiveContext, TInfo extends LiveInfo = AnyLiveInfo>
  implements LiveViewSocket<TContext, TInfo>
{
  abstract connected: boolean;
  abstract id: string;

  private _context: TContext;
  private _tempContext: Partial<TContext> = {}; // values to reset the context to post render cycle

  get context(): TContext {
    return structuredClone(this._context || ({} as TContext));
  }

  assign(context: Partial<TContext>) {
    this._context = {
      ...this.context,
      ...context,
    };
  }

  tempAssign(tempContext: Partial<TContext>) {
    this._tempContext = {
      ...this._tempContext,
      ...tempContext,
    };
  }

  pageTitle(newPageTitle: string) {
    // no-op
  }
  pushEvent(pushEvent: AnyLivePushEvent) {
    // no-op
  }
  pushPatch(path: string, params?: URLSearchParams, replaceHistory?: boolean) {
    // no-op
    // istanbul ignore next
    return Promise.resolve();
  }
  pushRedirect(path: string, params?: URLSearchParams, replaceHistory?: boolean) {
    // no-op
    // istanbul ignore next
    return Promise.resolve();
  }
  putFlash(key: string, value: string) {
    // no-op
    // istanbul ignore next
    return Promise.resolve();
  }
  repeat(fn: () => void, intervalMillis: number) {
    // no-op
  }
  sendInfo(info: Info<TInfo>) {
    // no-op
  }
  subscribe(topic: string) {
    // no-op
    // istanbul ignore next
    return Promise.resolve();
  }
  allowUpload(name: string, options?: UploadConfigOptions): Promise<void> {
    // no-op
    // istanbul ignore next
    return Promise.resolve();
  }
  cancelUpload(configName: string, ref: string): Promise<void> {
    // no-op
    // istanbul ignore next
    return Promise.resolve();
  }
  consumeUploadedEntries<T>(
    configName: string,
    fn: (meta: ConsumeUploadedEntriesMeta, entry: UploadEntry) => Promise<T>
  ): Promise<T[]> {
    // no-op
    // istanbul ignore next
    return Promise.resolve([]);
  }
  uploadedEntries(configName: string): Promise<{ completed: UploadEntry[]; inProgress: UploadEntry[] }> {
    // no-op
    // istanbul ignore next
    return Promise.resolve({ completed: [], inProgress: [] });
  }

  updateContextWithTempAssigns() {
    if (Object.keys(this._tempContext).length > 0) {
      this.assign(this._tempContext);
    }
  }
}

/**
 * Used to render Http requests for `LiveView`s.  Only support setting the context via
 * `assign` and reading the context via `context`.
 */
export class HttpLiveViewSocket<
  TContext extends LiveContext = AnyLiveContext,
  TInfo extends LiveInfo = AnyLiveInfo
> extends BaseLiveViewSocket<TContext, TInfo> {
  readonly id: string;
  readonly connected: boolean = false;
  readonly uploadConfigs: { [name: string]: UploadConfig } = {};

  private _redirect: { to: string; replace: boolean } | undefined;
  constructor(id: string) {
    super();
    this.id = id;
  }

  get redirect(): { to: string; replace: boolean } | undefined {
    return this._redirect;
  }

  pushRedirect(path: string, params?: URLSearchParams, replaceHistory?: boolean) {
    const to = params ? `${path}?${params}` : path;
    this._redirect = {
      to,
      replace: replaceHistory || false,
    };
    return Promise.resolve();
  }

  allowUpload(name: string, options?: UploadConfigOptions | undefined): Promise<void> {
    this.uploadConfigs[name] = new UploadConfig(name, options);
    return Promise.resolve();
  }
}

/**
 * Full inmplementation used once a `LiveView` is mounted to a websocket.
 * In practice, this uses callbacks defined in the LiveViewManager that
 * capture state and manipulate it in the context of that LiveViewManager instance.
 */
export class WsLiveViewSocket<
  TContext extends LiveContext = AnyLiveContext,
  TInfo extends LiveInfo = AnyLiveInfo
> extends BaseLiveViewSocket<TContext, TInfo> {
  readonly id: string;
  readonly connected: boolean = true;

  // callbacks to the ComponentManager
  private pageTitleCallback: (newPageTitle: string) => void;
  private pushEventCallback: (pushEvent: AnyLivePushEvent) => void;
  private pushPatchCallback: (path: string, params?: URLSearchParams, replaceHistory?: boolean) => void;
  private pushRedirectCallback: (path: string, params?: URLSearchParams, replaceHistory?: boolean) => void;
  private putFlashCallback: (key: string, value: string) => void;
  private repeatCallback: (fn: () => void, intervalMillis: number) => void;
  private sendInfoCallback: (info: Info<TInfo>) => void;
  private subscribeCallback: (topic: string) => void;
  private allowUploadCallback: (name: string, options?: UploadConfigOptions) => void;
  private cancelUploadCallback: (configName: string, ref: string) => void;
  private consumeUploadedEntriesCallback: <T>(
    configName: string,
    fn: (meta: ConsumeUploadedEntriesMeta, entry: UploadEntry) => Promise<T>
  ) => Promise<T[]>;
  private uploadedEntriesCallback: (
    configName: string
  ) => Promise<{ completed: UploadEntry[]; inProgress: UploadEntry[] }>;

  constructor(
    id: string,
    pageTitleCallback: (newPageTitle: string) => void,
    pushEventCallback: (pushEvent: AnyLivePushEvent) => void,
    pushPatchCallback: (path: string, params?: URLSearchParams, replaceHistory?: boolean) => void,
    pushRedirectCallback: (path: string, params?: URLSearchParams, replaceHistory?: boolean) => void,
    putFlashCallback: (key: string, value: string) => void,
    repeatCallback: (fn: () => void, intervalMillis: number) => void,
    sendInfoCallback: (info: Info<TInfo>) => void,
    subscribeCallback: (topic: string) => void,
    allowUploadCallback: (name: string, options?: UploadConfigOptions) => void,
    cancelUploadCallback: (configName: string, ref: string) => void,
    consumeUploadedEntriesCallback: <T>(
      configName: string,
      fn: (meta: ConsumeUploadedEntriesMeta, entry: UploadEntry) => Promise<T>
    ) => Promise<T[]>,
    uploadedEntriesCallback: (configName: string) => Promise<{ completed: UploadEntry[]; inProgress: UploadEntry[] }>
  ) {
    super();
    this.id = id;
    this.pageTitleCallback = pageTitleCallback;
    this.pushEventCallback = pushEventCallback;
    this.pushPatchCallback = pushPatchCallback;
    this.pushRedirectCallback = pushRedirectCallback;
    this.putFlashCallback = putFlashCallback;
    this.repeatCallback = repeatCallback;
    this.sendInfoCallback = sendInfoCallback;
    this.subscribeCallback = subscribeCallback;
    this.allowUploadCallback = allowUploadCallback;
    this.cancelUploadCallback = cancelUploadCallback;
    this.consumeUploadedEntriesCallback = consumeUploadedEntriesCallback;
    this.uploadedEntriesCallback = uploadedEntriesCallback;
  }
  async putFlash(key: string, value: string) {
    await this.putFlashCallback(key, value);
  }
  pageTitle(newPageTitle: string) {
    this.pageTitleCallback(newPageTitle);
  }
  pushEvent(pushEvent: AnyLivePushEvent) {
    this.pushEventCallback(pushEvent);
  }
  async pushPatch(path: string, params?: URLSearchParams, replaceHistory: boolean = false) {
    await this.pushPatchCallback(path, params, replaceHistory);
  }
  async pushRedirect(path: string, params?: URLSearchParams, replaceHistory: boolean = false) {
    await this.pushRedirectCallback(path, params, replaceHistory);
  }
  repeat(fn: () => void, intervalMillis: number) {
    this.repeatCallback(fn, intervalMillis);
  }
  sendInfo(info: Info<TInfo>): void {
    this.sendInfoCallback(info);
  }
  async subscribe(topic: string) {
    await this.subscribeCallback(topic);
  }
  async allowUpload(name: string, options?: UploadConfigOptions | undefined): Promise<void> {
    await this.allowUploadCallback(name, options);
  }
  async cancelUpload(configName: string, ref: string): Promise<void> {
    await this.cancelUploadCallback(configName, ref);
  }
  async consumeUploadedEntries<T>(
    configName: string,
    fn: (meta: ConsumeUploadedEntriesMeta, entry: UploadEntry) => Promise<T>
  ): Promise<T[]> {
    return await this.consumeUploadedEntriesCallback<T>(configName, fn);
  }
  async uploadedEntries(configName: string): Promise<{ completed: UploadEntry[]; inProgress: UploadEntry[] }> {
    return await this.uploadedEntriesCallback(configName);
  }
}
