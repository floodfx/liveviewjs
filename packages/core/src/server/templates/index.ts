export * from "./diff";
export * from "./helpers";
export * from "./htmlSafeString";
export * from "./jsCommands";
