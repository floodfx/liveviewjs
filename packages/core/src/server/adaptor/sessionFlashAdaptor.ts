import { SessionData } from "../session";
import { FlashAdaptor } from "./flash";

/**
 * Naive implementation of flash adaptor that uses "__flash" property on session data
 * to implement flash.
 */
export class SessionFlashAdaptor implements FlashAdaptor {
  peekFlash(session: SessionData, key: string): Promise<string | undefined> {
    if (!session.__flash) {
      // istanbul ignore next
      session.__flash = {};
    }
    return Promise.resolve(session.__flash[key]);
  }

  popFlash(session: SessionData, key: string): Promise<string | undefined> {
    // istanbul ignore next
    if (session.__flash === undefined) {
      // istanbul ignore next
      session.__flash = {};
    }
    const value = session.__flash[key];
    delete session.__flash[key];
    return Promise.resolve(value);
  }

  putFlash(session: SessionData, key: string, value: string): Promise<void> {
    if (!session.__flash) {
      // istanbul ignore next
      session.__flash = {};
    }
    session.__flash[key] = value;
    return Promise.resolve();
  }

  clearFlash(session: SessionData, key: string): Promise<void> {
    // istanbul ignore next
    if (session.__flash === undefined) {
      // istanbul ignore next
      session.__flash = {};
    }
    delete session.__flash[key];
    return Promise.resolve();
  }
}
