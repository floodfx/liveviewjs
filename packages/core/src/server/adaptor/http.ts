import {
  AnyLiveContext,
  HttpLiveComponentSocket,
  LiveComponent,
  LiveView,
  LiveViewHtmlPageTemplate,
  LiveViewTemplate,
  LiveViewWrapperTemplate,
  PathParams,
} from "../live";
import { SessionData } from "../session";
import { HttpLiveViewSocket } from "../socket/liveSocket";
import { html, safe } from "../templates";
import { LiveTitleOptions } from "../templates/helpers/live_title";
import { CsrfGenerator } from "./csrfGen";
import { IdGenerator } from "./idGen";
import { SerDe } from "./serDe";

/**
 * An interface that represents how to extract required data from an HTTP server request (such as Express, Koa, etc in the
 * Node ecosystem or Oak on the Deno ecosystem) for handling a LiveView http request.
 */
export interface HttpRequestAdaptor {
  /**
   * Make available SessionData from the HTTP request
   */
  getSessionData: () => SessionData;
  /**
   * Expose the HTTP request URL
   */
  getRequestUrl: () => URL;
  /**
   * Expose the path of the HTTP request URL
   */
  getRequestPath: () => string;
  /**
   * Callback to the adaptor that a redirect should be perfermed with the given url
   */
  onRedirect: (toUrl: string) => void;
  /**
   * Make available the SerDe for serializing and deserializing session data.
   */
  getSerDe: () => SerDe;
}

/**
 * Use the given inputs to handle (e.g. generate the HTML) for the requested LiveView. Usually this
 * is called via HTTP server middleware that determines if a request is to a LiveView and if so,
 * creates and passes all of the required inputs to this function
 * @param idGenerator the @{IdGenerator} to use to generate unique IDs for the HTTP request and web socket connection
 * @param csrfGenerator the @{CsrfGenerator} to use to generate unique CSRF tokens to protect against CSRF attacks
 * @param liveView the @{LiveView} to render
 * @param adaptor the @{HttpRequestAdaptor} to use to extract required data from the HTTP request
 * @param rootTemplateRenderer the @{LiveViewTemplate} which this @{LiveView} is rendered within (typically reused across all LiveViews)
 * @param pageTitleDefaults optional @{PageTitleDefaults} to use to set the page title for the LiveView
 * @param liveViewTemplateRenderer optional @{LiveViewTemplate} used for adding additional content to the LiveView (typically reused across all LiveViews)
 * @returns the HTML for the HTTP server to return to the client
 */
export const handleHttpLiveView = async (
  idGenerator: IdGenerator,
  csrfGenerator: CsrfGenerator,
  liveView: LiveView,
  adaptor: HttpRequestAdaptor,
  pageRenderer: LiveViewHtmlPageTemplate,
  pathParams: PathParams,
  pageTitleDefaults?: LiveTitleOptions,
  rootRenderer?: LiveViewWrapperTemplate
) => {
  const { getSessionData, getRequestUrl, onRedirect } = adaptor;
  // new LiveViewId for each request
  const liveViewId = idGenerator();

  // extract csrf token from session data or generate it if it doesn't exist
  const sessionData = getSessionData();
  if (sessionData._csrf_token === undefined) {
    sessionData._csrf_token = csrfGenerator();
  }

  // prepare a http socket for the `LiveView` render lifecycle: mount => handleParams => render
  const liveViewSocket = new HttpLiveViewSocket<AnyLiveContext>(liveViewId);

  // execute the `LiveView`'s `mount` function, passing in the data from the HTTP request
  await liveView.mount(
    liveViewSocket,
    { ...sessionData },
    { _csrf_token: sessionData._csrf_token, _mounts: -1, ...pathParams }
  );

  // check for redirects in `mount`
  if (liveViewSocket.redirect) {
    const { to } = liveViewSocket.redirect;
    onRedirect(to);
    return;
  }

  // execute the `LiveView`'s `handleParams` function, passing in the data from the HTTP request
  const url = getRequestUrl();
  await liveView.handleParams(url, liveViewSocket);

  // check for redirects in `handleParams`
  if (liveViewSocket.redirect) {
    const { to } = liveViewSocket.redirect;
    onRedirect(to);
    return;
  }

  // now render the `LiveView` including running the lifecycle of any `LiveComponent`s it contains
  let myself = 1; // counter for live_component calls
  const view = await liveView.render(liveViewSocket.context, {
    csrfToken: sessionData.csrfToken,
    async live_component(
      liveComponent: LiveComponent,
      params?: Partial<unknown & { id: string | number }>
    ): Promise<LiveViewTemplate> {
      // params may be empty if the `LiveComponent` doesn't have any params
      params = params ?? {};
      delete params.id; // remove id before passing to socket

      // prepare a http socket for the `LiveComponent` render lifecycle: mount => update => render
      const lcSocket = new HttpLiveComponentSocket(liveViewId, params);

      // pass params provided in `LiveView.render` to the `LiveComponent` socket
      lcSocket.assign(params);

      // start the `LiveComponent` lifecycle
      await liveComponent.mount(lcSocket);
      await liveComponent.update(lcSocket);

      // render view with context
      const newView = await liveComponent.render(lcSocket.context, { myself: myself });
      myself++;
      // return the view to the parent `LiveView` to be rendered
      return newView;
    },
    url,
    uploads: liveViewSocket.uploadConfigs,
  });

  // now that we've rendered the `LiveView` and its `LiveComponent`s, we can serialize the session data
  // to be passed into the websocket connection
  const serDe = adaptor.getSerDe();
  const serializedSession = await serDe.serialize({ ...sessionData });

  // TODO implement tracking of statics
  // const serializedStatics = serDe.serialize({ ...view.statics });
  const serializedStatics = "";

  // optionally render the `LiveView` inside another template passing the session data
  // and the rendered `LiveView` to the template renderer
  let liveViewContent = safe(view);
  if (rootRenderer) {
    liveViewContent = await rootRenderer({ ...sessionData }, safe(view));
  }

  // wrap `LiveView` content inside the `phx-main` template along with the serialized
  // session data and the generated live view ID for the websocket connection
  const rootContent = html`
    <div
      data-phx-main="true"
      data-phx-session="${serializedSession}"
      data-phx-static="${serializedStatics}"
      id="phx-${liveViewId}">
      ${safe(liveViewContent)}
    </div>
  `;

  // finally render the `LiveView` root template passing any pageTitle data, the CSRF token,  and the rendered `LiveView`
  const rootView = await pageRenderer(pageTitleDefaults ?? { title: "" }, sessionData._csrf_token, rootContent);
  return rootView.toString();
};
