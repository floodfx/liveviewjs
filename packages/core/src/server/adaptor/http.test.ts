import {
  AnyLiveContext,
  BaseLiveComponent,
  BaseLiveView,
  createLiveView,
  LiveComponentMeta,
  LiveViewMeta,
  LiveViewMountParams,
  LiveViewTemplate,
} from "../live";
import { SessionData } from "../session";
import { LiveViewSocket } from "../socket";
import { html } from "../templates";
import { CsrfGenerator } from "./csrfGen";
import { handleHttpLiveView, HttpRequestAdaptor } from "./http";
import { IdGenerator } from "./idGen";
import { JsonSerDe } from "./jsonSerDe";
import { SessionFlashAdaptor } from "./sessionFlashAdaptor";

const idGen: IdGenerator = (id?: string) => id || "id";
const csrfGen: CsrfGenerator = (csrf?: string) => csrf || "csrf";

describe("test http adaptor", () => {
  it("adapts to http live views", async () => {
    const view = await handleHttpLiveView(
      idGen,
      csrfGen,
      new TestLiveView(),
      new TestHttpAdaptor(),
      (pageTitleDefault, csrfToken, content) => {
        return html`<main>${content}</main>`;
      },
      {}
    );

    expect(view).toMatchSnapshot();
  });

  it("handles redirect in mount", async () => {
    const redirectM = true;
    const adaptor = new TestHttpAdaptor();
    const view = await handleHttpLiveView(
      idGen,
      csrfGen,
      new TestLiveView(redirectM),
      adaptor,
      (pageTitleDefault, csrfToken, content) => {
        return html`<main>${content}</main>`;
      },
      {}
    );

    expect(adaptor.redirect).toEqual("/redirectM");
    expect(view).toBeUndefined();
  });

  it("handles redirect in HP", async () => {
    const redirectM = false;
    const redirectHP = true;
    const adaptor = new TestHttpAdaptor();
    const view = await handleHttpLiveView(
      idGen,
      csrfGen,
      new TestLiveView(redirectM, redirectHP),
      adaptor,
      (pageTitleDefault, csrfToken, content) => {
        return html`<main>${content}</main>`;
      },
      {}
    );

    expect(adaptor.redirect).toEqual("/redirectHP");
    expect(view).toBeUndefined();
  });

  it("shows live component", async () => {
    const redirectM = false;
    const redirectHP = false;
    const showLiveComponent = true;
    const adaptor = new TestHttpAdaptor();
    const view = await handleHttpLiveView(
      idGen,
      csrfGen,
      new TestLiveView(redirectM, redirectHP, showLiveComponent),
      adaptor,
      (pageTitleDefault, csrfToken, content) => {
        return html`<main>${content}</main>`;
      },
      {}
    );

    expect(view).toMatchSnapshot();
  });

  it("shows live stateless component", async () => {
    const redirectM = false;
    const redirectHP = false;
    const showLiveComponent = true;
    const adaptor = new TestHttpAdaptor();
    const view = await handleHttpLiveView(
      idGen,
      csrfGen,
      testEmptyParamLC,
      adaptor,
      (pageTitleDefault, csrfToken, content) => {
        return html`<main>${content}</main>`;
      },
      {}
    );

    expect(view).toMatchSnapshot();
  });

  it("rendering view inside another template", async () => {
    const redirectM = false;
    const redirectHP = false;
    const showLiveComponent = true;
    const adaptor = new TestHttpAdaptor();
    const view = await handleHttpLiveView(
      idGen,
      csrfGen,
      new TestLiveView(redirectM, redirectHP, showLiveComponent),
      adaptor,
      (pageTitleDefault, csrfToken, content) => {
        return html`<main>${content}</main>`;
      },
      {},
      { title: "Default Title" },
      async (session, content) => {
        const flashAdaptor = new SessionFlashAdaptor();
        const infoFlash = (await flashAdaptor.popFlash(session, "test")) || "";
        return html`<div class="container">
          <p class="flash">test: ${infoFlash}</p>
          ${content}
        </div>`;
      }
    );

    expect(view).toMatchSnapshot();
  });
});

class TestHttpAdaptor implements HttpRequestAdaptor {
  redirect: string;
  getRequestPath = () => {
    return "/test";
  };
  getRequestUrl = () => {
    return new URL("http://example.com/test");
  };
  onRedirect = (toUrl: string) => {
    this.redirect = toUrl;
  };
  getSerDe = () => {
    return new JsonSerDe();
  };
  getSessionData = () => {
    const session: SessionData = {};
    const flashAdaptor = new SessionFlashAdaptor();
    flashAdaptor.putFlash(session, "test", "test message");
    return session;
  };
}

class TestLiveComponent extends BaseLiveComponent {
  render(context: AnyLiveContext, meta: LiveComponentMeta): LiveViewTemplate | Promise<LiveViewTemplate> {
    return html`<div>Test Live Component: ${meta.myself}</div>`;
  }
}

class TestLiveView extends BaseLiveView {
  redirectM?: boolean;
  redirectHP?: boolean;
  showLiveComponent?: boolean;
  constructor(redirectM: boolean = false, redirectHP: boolean = false, showLiveComponent: boolean = false) {
    super();
    this.redirectM = redirectM;
    this.redirectHP = redirectHP;
    this.showLiveComponent = showLiveComponent;
  }
  mount(socket: LiveViewSocket<AnyLiveContext>, session: Partial<SessionData>, params: LiveViewMountParams): void {
    if (this.redirectM) {
      socket.pushRedirect("/redirectM");
    }
  }
  handleParams(url: URL, socket: LiveViewSocket<AnyLiveContext>): void {
    if (this.redirectHP) {
      socket.pushRedirect("/redirectHP");
    }
  }
  async render(context: AnyLiveContext, meta: LiveViewMeta): Promise<LiveViewTemplate> {
    // console.log("showLiveComponent", this.showLiveComponent);
    return html`
      <div>
        Test Live View
        ${this.showLiveComponent
          ? html`<div>${await meta.live_component(new TestLiveComponent(), { id: 1 })}</div>`
          : ""}
      </div>
    `;
  }
}

const testEmptyParamLC = createLiveView({
  render: async (ctx, meta) => {
    return html`<div>${await meta.live_component(new TestLiveComponent())}</div>`;
  },
});
