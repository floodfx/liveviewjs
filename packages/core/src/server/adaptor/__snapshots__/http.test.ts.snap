// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test http adaptor adapts to http live views 1`] = `
"<main>
    <div
      data-phx-main=\\"true\\"
      data-phx-session=\\"{&quot;__flash&quot;:{&quot;test&quot;:&quot;test message&quot;},&quot;_csrf_token&quot;:&quot;csrf&quot;}\\"
      data-phx-static=\\"\\"
      id=\\"phx-id\\">
      
      <div>
        Test Live View
        
      </div>
    
    </div>
  </main>"
`;

exports[`test http adaptor rendering view inside another template 1`] = `
"<main>
    <div
      data-phx-main=\\"true\\"
      data-phx-session=\\"{&quot;__flash&quot;:{&quot;test&quot;:&quot;test message&quot;},&quot;_csrf_token&quot;:&quot;csrf&quot;}\\"
      data-phx-static=\\"\\"
      id=\\"phx-id\\">
      <div class=\\"container\\">
          <p class=\\"flash\\">test: test message</p>
          
      <div>
        Test Live View
        <div><div>Test Live Component: 1</div></div>
      </div>
    
        </div>
    </div>
  </main>"
`;

exports[`test http adaptor shows live component 1`] = `
"<main>
    <div
      data-phx-main=\\"true\\"
      data-phx-session=\\"{&quot;__flash&quot;:{&quot;test&quot;:&quot;test message&quot;},&quot;_csrf_token&quot;:&quot;csrf&quot;}\\"
      data-phx-static=\\"\\"
      id=\\"phx-id\\">
      
      <div>
        Test Live View
        <div><div>Test Live Component: 1</div></div>
      </div>
    
    </div>
  </main>"
`;

exports[`test http adaptor shows live stateless component 1`] = `
"<main>
    <div
      data-phx-main=\\"true\\"
      data-phx-session=\\"{&quot;__flash&quot;:{&quot;test&quot;:&quot;test message&quot;},&quot;_csrf_token&quot;:&quot;csrf&quot;}\\"
      data-phx-static=\\"\\"
      id=\\"phx-id\\">
      <div><div>Test Live Component: 1</div></div>
    </div>
  </main>"
`;
